lib/analyzer/image-extractor.ts

```diff
@@ -1,110 +1,164 @@
 }
 
-interface Layers {
-  [key: string]: ExtractedKeyFiles;
+interface SearchAction {
+  name: string; // name, should be unique, for this action
+  pattern: string; // path pattern to look for
+  callback: SearchActionCallback; // convert Buffer into a string or Buffer
 }
 
 /**
  * Extract key files form the specified TAR stream.
- * @param stream image layer as a Readable TAR stream
- * @param name layer name
- * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
- * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
- * @returns key files found in the specified layer TAR stream
+ * @param layerTarStream image layer as a Readable TAR stream
+ * @param searchActions array of pattern, callbacks pairs
+ * @returns extracted file products
  */
-async function extractLayerKeyFiles(
-  stream: Readable,
-  txtPatterns: string[],
-  md5Patterns: string[] = [],
-): Promise<ExtractedKeyFiles> {
+async function extractFromLayer(
+  layerTarStream: Readable,
+  searchActions: SearchAction[],
+): Promise<SearchActionProducts> {
   return new Promise((resolve) => {
-    const result: ExtractedKeyFiles = { txt: {}, md5: {} };
+    const result: SearchActionProducts = {};
     const layerExtract: Extract = extract();
-    layerExtract.on("entry", (header, stream, next) => {
-      txtPatterns.forEach((pattern) => {
-        if (minimatch(header.name, pattern, { dot: true })) {
-          streamToString(stream).then((value) => {
-            result.txt[header.name] = value;
-          });
-        }
-      });
-      md5Patterns.forEach((pattern) => {
-        if (minimatch(header.name, pattern, { dot: true })) {
-          streamToBuffer(stream).then((value) => {
-            result.md5[header.name] = md5(value);
-          });
+    layerExtract.on("entry", async (header, stream, next) => {
+      if (
+        header.type === "file" ||
+        header.type === "link" ||
+        header.type === "symlink"
+      ) {
+        const filename = `/${header.name}`;
+        // convert stream to buffer in order to allow it
+        //  to be processed multiple times by the callback
+        const buffer = await streamToBuffer(stream);
+        for (const searchAction of searchActions) {
+          if (minimatch(filename, searchAction.pattern, { dot: true })) {
+            if (header.type === "file") {
+              // initialize the files associated products dict
+              if (!result[filename]) {
+                result[filename] = {};
+              }
+              try {
+                // store the product under the search action name
+                result[filename][searchAction.name] = searchAction.callback(
+                  buffer,
+                );
+              } catch (error) {
+                // prevent static scan from crash due to callback error
+                debug(`${filename} ${searchAction.name} ${error}`);
```

Comment:
I think we should throw in this case - this way, you'll not get familiar with the fact that there is any problem...